samcli/lib/test_runner/fargate_testsuite_runner.py

```diff
@@ -0,0 +1,472 @@
+                LOG.info(
+                    self.color.yellow(
+                        f"=> There are no updates to be performed on the stack '{self.runner_stack_name}', proceeding with testsuite execution...\n",
+                    )
+                )
+            else:
+                raise DeployFailedError(
+                    stack_name=self.runner_stack_name, msg=str(deployment_exception)
+                ) from deployment_exception
+
+    def _update_or_create_test_runner_stack(self):
+        """
+        Creates a new Test Runner stack from a given template if a stack with the given name does not yet exist, or updates the existing Test Runner stack.
+
+        Raises
+        ------
+        MissingTestRunnerTemplateException
+            If the customer specifies a Test Runner stack name that does not exist, and also fails to provide a template to create it.
+        """
+        runner_stack_exists = self.deployer.has_stack(stack_name=self.runner_stack_name)
+
+        if not runner_stack_exists:
+            if not self.runner_template_path:
+                raise MissingTestRunnerTemplateException(
+                    f"There does not exist a stack named '{self.runner_stack_name}'. Please provide a Test Runner CloudFormation template with `--runner-template`, and the stack will be created.\n"
+                    "To have a Test Runner CloudFormation template generated, run `sam test_runner init`.\n",
+                )
+            self._create_new_test_runner_stack(template_body=self._read_file(self.runner_template_path))
+            return
+
+        # Stack exists, but a template was also provided, so update
+        if self.runner_template_path:
+            self._update_exisiting_test_runner_stack(template_body=self._read_file(self.runner_template_path))
+
+    def _get_resource_list(self) -> List[dict]:
+        """
+        Returns a list of `StackResource` objects that compose the Test Runner Stack
+
+        NOTE: `StackResource` definition: https://awscli.amazonaws.com/v2/documentation/api/latest/reference/cloudformation/describe-stack-resources.html#output
+        """
+        stack_information = self.boto_cloudformation_client.describe_stack_resources(StackName=self.runner_stack_name)
+        resource_list = stack_information.get("StackResources")
+        return resource_list
+
+    def _get_unique_resource_physical_id(self, resource_list: List[dict], resource_type: str) -> str:
+        """
+        Extracts the physical ID of a resource with a given type from a given list of Test Runner `StackResources`
+
+        An exception is thrown if there is no resource with the given type present in `resource_list` OR if there is more than one.
+
+        This is because resources such as S3 Buckets and ECS Clusters are automatically picked up from the Test Runner Stack, and if there is more than one, the question of which to use is ambiguous.
```

Comment:
I don't suppose we could figure out if there are too many resources before we deploy the stack? I don't think this is a blocking issue, but it seems annoying to have to wait for a CFN deployment before finding out an otherwise correct change to the generated template is disallowed.

Reply:
We could parse through the template before deploying? I'll look into this refactor

Reply:
Given how small this problem is expected to be (I don't see why customers would go adding buckets or task definitions to the Test Runner template), I think we can stick with the reliability of the API calls instead of adding functionality to parse through the template.

If this is an issue in the future, we can easily remove the checks from the `FargateTestuiteRunner`, and then parse and check the template beforehand.